### PR TITLE
Upgrade babe to the latest

### PR DESCRIPTION
**Pull request by cowboy-bebug**
Here is the range of commits looked at:
- https://github.com/paritytech/substrate/commit/6ce091737dc7f114dd420f8b879f8a467c1d501b: start
- https://github.com/paritytech/substrate/commit/99274e39c715cb94a89a7164289a01031be15d38: end

Notable commits:
- https://github.com/paritytech/substrate/commit/d1238423a3093eb64df90c0eebdbf504c8df56dd
  - `EstimateNextSessionRotation` trait is added and implemented for `pallet-babe::Trait`
- https://github.com/paritytech/substrate/commit/2b67057a020c4973b3af68a6868b1aa2b982e03d
  - WIP
- https://github.com/paritytech/substrate/commit/18334ee173f0ec4b62d2ca05c5b0c1f0b05b1b17 - equivocation reporting 
  - will be looked at after the above

Commits not brought in:
- https://github.com/paritytech/substrate/commit/80e4417f1bbece16ec4c577d22dfb8c3e48a1375 - we don't use offence-benchmarks
- https://github.com/paritytech/substrate/commit/7fccdfac8232b2286c3dfd481ffcb2744ebc070b - we don't have a planned config update for Babe

Ignored weight-related commits:
- https://github.com/paritytech/substrate/commit/940e52110464cefaae4c1ab98db3cfaacf4d8287
- https://github.com/paritytech/substrate/commit/899a44864900cb691c99f7421bc85e4118bca93d
- https://github.com/paritytech/substrate/commit/e824e8ab0fadec9949ebb8b9e14d98703d6b8d44

Concern:
- sp_core::traits::SpawnNamed not brought in
- BasicQueue::new spawner param not brought in
- `frame_system::offchain::SendTransactionTypes` not brought in (can't send equivocation report yet)

### CODE REVIEW

File: frame/babe/src/mock.rs [hunk 1]
```diff
@@ -70,19 +81,40 @@ impl frame_system::Trait for Test {
...
 	}
 }
 
 impl pallet_session::Trait for Test {
 	type Event = ();
 	type ValidatorId = <Self as frame_system::Trait>::AccountId;
 	type ShouldEndSession = Babe;
-	type SessionHandler = (Babe,Babe,);
-	type SessionManager = ();
+	type SessionHandler = (Babe,);
```
Comment (jordy25519) [inline frame/babe/src/mock.rs]: [changed upstream](https://github.com/paritytech/substrate/commit/18334ee173f0ec4b62d2ca05c5b0c1f0b05b1b17#diff-77d405d7362bce705009a3f7e5bf95d0R116)
Is there a reason not to change to that?
```suggestion
	type SessionHandler = <MockSessionKeys as OpaqueKeys>::KeyTypeIdProviders;
```

File: frame/babe/src/mock.rs [hunk 2]
```diff
@@ -70,19 +81,40 @@ impl frame_system::Trait for Test {
...
+	type SessionManager = pallet_session::historical::NoteHistoricalRoot<Self, Staking>;
 	type ValidatorIdOf = ();
```
Comment (jordy25519) [inline frame/babe/src/mock.rs]: this one changed upstream also
```suggestion
	type ValidatorIdOf = pallet_staking::StashOf<Self>;
```

File: client/consensus/epochs/src/lib.rs [hunk 3]
```diff
@@ -335,6 +335,55 @@ impl<Hash, Number, E: Epoch> EpochChanges<Hash, Number, E> where
 		self.inner.rebalance()
 	}
 
+		/// Map the epoch changes from one storing data to a different one.
```
Comment (jordy25519) [inline client/consensus/epochs/src/lib.rs]: nit: whitespace

File: client/consensus/babe/src/lib.rs [hunk 4]
```diff
@@ -49,6 +49,11 @@
 //!
```
Comment (jordy25519) [inline client/consensus/babe/src/lib.rs:49]: Some changes in this file from: https://github.com/paritytech/substrate/pull/6008
Would be worth pulling in also (not sure if you tried)

File: client/consensus/babe/src/tests.rs [hunk 5]
```diff
@@ -22,7 +22,7 @@
 use super::*;
 use authorship::claim_slot;
```
Comment (jordy25519) [inline client/consensus/babe/src/tests.rs]: there's a test in this file which seems to be removed upstream: `babe_transcript_generation_match()` which could be removed

File: client/consensus/slots/src/lib.rs [hunk 6]
```diff
@@ -20,7 +20,6 @@
 //! time during which certain events can and/or must occur.  This crate
```
Comment (jordy25519) [inline client/consensus/slots/src/lib.rs:20]: some small changes related too: https://github.com/paritytech/substrate/pull/6008 which could be pulled in here also

File: frame/babe/src/equivocation.rs [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+	fn is_known_offence(offenders: &[T::KeyOwnerIdentification], time_slot: &SlotNumber) -> bool {
+		R::is_known_offence(offenders, time_slot)
+	}
+
+	fn submit_unsigned_equivocation_report(
+		equivocation_proof: EquivocationProof<T::Header>,
+		key_owner_proof: T::KeyOwnerProof,
+	) -> DispatchResult {
+		// TODO: bring in off-chain refactor around SendTransactionType
```
Comment (jordy25519) [inline frame/babe/src/equivocation.rs:147]: definitely needs a follow up PR

**Review by jordy25519 — APPROVED** [review]
looks great on the whole.
- definitely need to fix the equivocation transaction submission
- both [5776](https://github.com/paritytech/substrate/pull/5776) and [6008](https://github.com/paritytech/substrate/pull/6008) look useful and would bring us closer to upstream. could be easily added in a follow up PR.

File: client/consensus/babe/src/lib.rs (hunk 4 above)
Comment (cowboy-bebug) [reply client/consensus/babe/src/lib.rs:49]: Will bring it in a follow up PR.

File: client/consensus/slots/src/lib.rs (hunk 6 above)
Comment (cowboy-bebug) [reply client/consensus/slots/src/lib.rs:20]: Will bring this one in a follow up PR

**Pull request merged**